File: botasaurus_server/botasaurus_server/task_executor.py
from datetime import datetime, timezone
from threading import Thread, Lock
import time
from math import inf
from sqlalchemy import and_
import traceback
from .cleaners import clean_data
from .db_setup import Session
from .server import Server
from .models import Task, TaskStatus, remove_duplicates_by_key
from .task_helper import TaskHelper
from .scraper_type import ScraperType
from .retry_on_db_error import retry_on_db_error
from .task_results import TaskResults
from botasaurus.dontcache import is_dont_cache

class TaskExecutor:

    def load(self):
        self.current_capacity = {"browser": 0, "request": 0, "task": 0}
        self.lock = Lock()
    
    def start(self):
        self.fix_in_progress_tasks()
        Thread(target=self.task_worker, daemon=True).start()
    
    @retry_on_db_error
    def fix_in_progress_tasks(self):
        with Session() as session:
            # Delete tasks with is_sync=True and status as either pending or in progress
            session.query(Task).filter(
                and_(
                    Task.is_sync == True,
                    Task.status.in_([TaskStatus.PENDING, TaskStatus.IN_PROGRESS]),
                )
            ).delete()
            # Update in progress tasks to pending
            session.query(Task).filter(Task.status == TaskStatus.IN_PROGRESS).update(
                {"status": TaskStatus.PENDING, "started_at": None},
            )
            session.commit()


    def task_worker(self):
        browser_scrapers = len(Server.get_browser_scrapers()) > 0
        request_scrapers = len(Server.get_request_scrapers()) > 0
        task_scrapers = len(Server.get_task_scrapers()) > 0
        
        while True:
            try:
                if browser_scrapers:
                    self.process_tasks(ScraperType.BROWSER)

                if request_scrapers:
                    self.process_tasks(ScraperType.REQUEST)

                if task_scrapers:
                    self.process_tasks(ScraperType.TASK)
            except:
              traceback.print_exc()

            time.sleep(1)

    def process_tasks(self, scraper_type):
        rate_limit = self.get_max_running_count(scraper_type)
        current_count = self.get_current_running_count(scraper_type)
        if current_count < rate_limit:
            self.execute_pending_tasks(
                and_(
                    Task.status == TaskStatus.PENDING,
                    Task.scraper_type == scraper_type,
                    Task.is_all_task == False,
                ),
                current_count,
                rate_limit,
                scraper_type
            )

    def get_max_running_count(self, scraper_type):
        return Server.get_rate_limit()[scraper_type]

    def get_current_running_count(self, scraper_type):
        return self.current_capacity[scraper_type]

    @retry_on_db_error
    def execute_pending_tasks(self, task_filter, current, limit, scraper_type):
            with Session() as session:
                query = (
                    session.query(Task).filter(task_filter).order_by(Task.sort_id.desc(), Task.is_sync.desc()) # Prioritize syncronous tasks
                )
                if limit != inf:
                    remaining = limit - current
                    query = query.limit(remaining)

                tasks = query.all()

                if tasks:
                    # Collect task and parent IDs
                    task_ids = [task.id for task in tasks]
                    parent_ids = {task.parent_task_id for task in tasks if task.parent_task_id}

                    # Bulk update the status of tasks
                    if task_ids:
                        session.query(Task).filter(Task.id.in_(task_ids)).update(
                            {"status": TaskStatus.IN_PROGRESS, "started_at": datetime.now(timezone.utc)}
                        )

                    # Bulk update the status of parent tasks
                    if parent_ids:
                        session.query(Task).filter(
                            Task.id.in_(list(parent_ids)), Task.started_at.is_(None)
                        ).update(
                            {"status": TaskStatus.IN_PROGRESS, "started_at": datetime.now(timezone.utc)}
                        )

                    session.commit()

                    # Start tasks after commit
                    for task in tasks:
                        self.run_task_and_update_state(scraper_type, task.to_json())

    def run_task_and_update_state(self, scraper_type, task_json):
        Thread(target=self.run_task, args=(task_json,), daemon=True).start()
        self.increment_capacity(scraper_type)
        

    def increment_capacity(self, scraper_type):
        # Temp Lock Disabling to check if it reduces the deadlock on vm tasks
        # with self.lock:
            self.current_capacity[scraper_type] += 1

    def decrement_capcity(self, scraper_type):
        # Temp Lock Disabling to check if it reduces the deadlock on vm tasks
        # with self.lock:
            self.current_capacity[scraper_type] -= 1

    def run_task(self, task):

        scraper_type = task["scraper_type"]
        task_id = task["id"]
        scraper_name = task["scraper_name"]
        parent_task_id = task["parent_task_id"]
        metadata = {"metadata": task["metadata"]} if task["metadata"] != {} else {}
        task_data = task["data"]

        fn = Server.get_scraping_function(scraper_name)
        exception_log = None
        try:
            try:
                result = fn(
                    task_data,
                    **metadata,
                    parallel=None,
                    cache=False,
                    beep=False,
                    run_async=False,
                    async_queue=False,
                    raise_exception=True,
                    close_on_crash=True,
                    output=None,
                    create_error_logs=False,
                    return_dont_cache_as_is=True
                )
                if is_dont_cache(result):
                    is_result_dont_cached = True
                    # unpack
                    result = result.data
                else: 
                    is_result_dont_cached = False

                result = clean_data(result)

                remove_duplicates_by = Server.get_remove_duplicates_by(scraper_name)
                if remove_duplicates_by:
                   result = remove_duplicates_by_key(result, remove_duplicates_by)

                if is_result_dont_cached:
                    self.mark_task_as_success(task_id, result, False, scraper_name, task_data)
                else:
                    self.mark_task_as_success(task_id, result, Server.cache, scraper_name, task_data)
                self.decrement_capcity(scraper_type)
            except:
                self.decrement_capcity(scraper_type)
                exception_log = traceback.format_exc()
                traceback.print_exc()
                self.mark_task_as_failure(task_id, exception_log)
            finally:
                if parent_task_id:
                    if exception_log:
                        self.update_parent_task(task_id, [])
                    else:
                        self.update_parent_task(task_id, result)
        except Exception as e:
          traceback.print_exc()
          print("Error in run_task ", e)
    
    @retry_on_db_error
    def update_parent_task(self, task_id, result):
        with Session() as session:
            task_to_update = session.get(Task, task_id)
            parent_id = task_to_update.parent_task_id
            scraper_name = task_to_update.scraper_name
        
        if parent_id:
            self.complete_parent_task_if_possible(parent_id, Server.get_remove_duplicates_by(scraper_name), result)

    def complete_parent_task_if_possible(self,parent_id, remove_duplicates_by, result):
            fn = None
            with Session() as session:
                parent_task = TaskHelper.get_task(
                            session,
                            parent_id,
                            [TaskStatus.PENDING, TaskStatus.IN_PROGRESS],
                        )

                # MAY BE DELETED SO CHECK
                if parent_task:
                    # is fast no worry
                    TaskHelper.update_parent_task_results(parent_id,result,)
                    is_done = TaskHelper.are_all_child_task_done(
                                session, parent_id
                            )
                    if is_done:
                        failed_children_count = TaskHelper.get_failed_children_count(session, parent_id)
                        
                        status = TaskStatus.FAILED if failed_children_count else TaskStatus.COMPLETED
                        fn = lambda: TaskHelper.read_clean_save_task(parent_id, remove_duplicates_by, status)
                        
            if fn:
                fn()          

    def complete_as_much_all_task_as_possible(self,parent_id, remove_duplicates_by):
            fn = None
            with Session() as session:
                    is_done = TaskHelper.are_all_child_task_done(
                                session, parent_id
                            )
                    if is_done:
                        failed_children_count = TaskHelper.get_failed_children_count(session, parent_id)
                        status = TaskStatus.FAILED if failed_children_count else TaskStatus.COMPLETED
                        fn = lambda: TaskHelper.collect_and_save_all_task(parent_id, None, remove_duplicates_by, status)
                    else:
                        fn = lambda: TaskHelper.collect_and_save_all_task(parent_id, None, remove_duplicates_by, TaskStatus.IN_PROGRESS)
            if fn:
                fn()  

    @retry_on_db_error
    def mark_task_as_failure(self, task_id, exception_log):
        TaskResults.save_task(task_id, exception_log)
        with Session() as session:
            TaskHelper.update_task(
                    session,
                    task_id,
                    {
                        "status": TaskStatus.FAILED,
                        "finished_at": datetime.now(timezone.utc),
                    },
                    [TaskStatus.IN_PROGRESS],
                )
            session.commit()
    
    @retry_on_db_error
    def mark_task_as_success(self, task_id, result, cache_task, scraper_name, data):
        TaskResults.save_task(task_id, result)
        if cache_task:
            TaskResults.save_cached_task(scraper_name, data, result)
        with Session() as session:

            update_result = TaskHelper.update_task(
                    session,
                    task_id,
                    {
                        "result_count": len(result),
                        "status": TaskStatus.COMPLETED,
                        "finished_at": datetime.now(timezone.utc),
                    },
                    [TaskStatus.IN_PROGRESS],
                )
            if not update_result:
                # if the task is aborted in progress, there should be no results
                TaskResults.delete_task(task_id)

            session.commit()

File: js/botasaurus-server-js/src/task-executor.ts
import { db, serializeTaskForRunTask } from './models'
import { isNotNullish, isNullish } from './null-utils'
import { Task, TaskStatus, removeDuplicatesByKey } from './models'
import { TaskHelper, PushDataWriter } from './task-helper'
import { ScraperType } from './scraper-type'
import { TaskResults } from './task-results'
import { isDontCache } from 'botasaurus/dontcache'
import { formatExc } from 'botasaurus/utils'
import { cleanDataInPlace } from 'botasaurus/output'
import { getScraperErrorMessage, Server } from './server'
import { isLargeFile, isNotEmptyObject } from './utils'

/**
 * Task priority levels for queue ordering
 */
export enum TaskPriority {
    DEFAULT_PRIORITY = 0,
    URGENT_PRIORITY = 1
}
export function extractParentIds(tasks: any[]) {
    const parentIdsSet = new Set<any>();
    tasks.forEach((task) => {
        if (task.parent_task_id) {
            parentIdsSet.add(task.parent_task_id);
        }
    });
    return Array.from(parentIdsSet);
}

async function queryPendingTasks(maxTasks: number | null, query: any) {
    if (isNotNullish(maxTasks)) {
        // @ts-ignore
        return db.findAsync(query).sort({ priority: -1, sort_id: -1 }).limit(maxTasks) as unknown as Task[]

    } else {
        return await db.findAsync(query).sort({ priority: -1, sort_id: -1 }) as unknown as Task[]

    }
}


export async function getPendingTasks(query: any, maxTasks: number | null) {
    const tasks = await queryPendingTasks(maxTasks, query)
    
            if (tasks.length === 0) {
                return [];
            }

            const validScraperNames = Server.getScrapersNames()
            const validScraperNamesSet = new Set(validScraperNames)

            for (const task of tasks) {
                if (!validScraperNamesSet.has(task.scraper_name)) {
                    const validNamesString = validScraperNames.join(', ')
                    throw new Error(getScraperErrorMessage(validScraperNames, task.scraper_name, validNamesString))
                }
            }            

            const taskIds = tasks.map(t => t.id);
            const parentIds = extractParentIds(tasks)

            // Mark tasks as IN_PROGRESS
            await db.updateAsync(
                { id: { $in: taskIds } },
                {
                    $set: {
                        status: TaskStatus.IN_PROGRESS,
                        started_at: new Date()
                    }
                },
                { multi: true }
            );



            // Update parent tasks if applicable
            if (parentIds.length > 0) {
                try {
                    await db.updateAsync(
                        { id: { $in: parentIds }, started_at: null },
                        { $set: { status: TaskStatus.IN_PROGRESS, started_at: new Date() } },
                        { multi: true }
                    );
                } catch (parentErr) {
                    console.error('Error updating parent tasks:', parentErr);
                }
            }

            // Serialize tasks for workers
            return tasks.map(task => serializeTaskForRunTask(task));
}

class TaskExecutor {
    currentCapacity: { browser?: number; request?: number; task?: number } | Record<string, number> = {
        browser: 0,
        request: 0,
        task: 0,
    }

    protected readonly ABORT_CHECK_INTERVAL: number = 1000
    
    // Centralized abortion checking state
    private toCheckAbortTasks: Set<number> = new Set()
    private abortedTasksState: Record<number, boolean> = {}
    private abortCheckIntervalId: ReturnType<typeof setInterval> | null = null

    public load() {
        if (Server.isScraperBasedRateLimit) {
            this.currentCapacity = {}
        } else {
            this.currentCapacity = { browser: 0, request: 0, task: 0 }
        }

        Server.validateRateLimit()
    }

    public async start(): Promise<void> {
        await this.fixInProgressTasks()
        await this.completePendingButCompletedAllTask()
        this.startTaskWorker()
    }

    protected startTaskWorker(): void {
        if (Server.isScraperBasedRateLimit) {
            setImmediate(this.taskWorkerScraperBased.bind(this))
        } else {
            setImmediate(this.taskWorkerScraperTypeBased.bind(this))
        }
    }

    private async completePendingButCompletedAllTask(): Promise<void> {
        try {
            const pendingTasks = await new Promise<Task[]>((resolve, reject) => {
                db.find(
                  {
                    $and: [
                      { is_all_task: true },
                      { status: TaskStatus.PENDING },
                    ],
                  },
                  (err: any, tasks: Task[]) => {
                    if (err) {
                      reject(err)
                    } else {
                      resolve(tasks)
                    }
                  }
                )
              })
          
              for (const parentTask of pendingTasks) {
                const parentId = parentTask.id
                const isAllChildTasksDone = await TaskHelper.areAllChildTaskDone(parentId)
                if (isAllChildTasksDone) {
                  const is_large: any = parentTask.is_large
      
                  const failedChildrenCount = await TaskHelper.getFailedChildrenCount(parentId)
                  const status = failedChildrenCount ? TaskStatus.FAILED : TaskStatus.COMPLETED
                  
                  const removeDuplicatesBy = Server.getRemoveDuplicatesBy(parentTask.scraper_name)
                  // Find the oldest started_at task that is the parent of the current task
                  const oldestParentTask = await this.findOldestChildTask(parentId)
                  
                  const extra_updates = (finished_at: any) => ({
                      started_at: oldestParentTask ? oldestParentTask.started_at : finished_at,
                  })
                  
                  await TaskHelper.readCleanSaveTask(
                      parentId,
                      removeDuplicatesBy,
                      status,
                      is_large, 
                      extra_updates
                  )
                }
              }
        } catch (error) {
            console.error(error)
        }
    }
    
    private async findOldestChildTask(parentId: number) {
        return await new Promise<Task | null>((resolve, reject) => {
            db.find({
                $and: [
                    { parent_task_id: parentId },
                    { started_at: { $exists: true } },
                ],
            })
            .sort({ started_at: 1 }) // oldest
            .limit(1)
            .exec((err: any, tasks: any[]) => {
                if (err) {
                    reject(err)
                } else {
                    resolve(tasks.length > 0 ? tasks[0] : null)
                }
            })
        })
    }

    private async fixInProgressTasks(): Promise<void> {
        try {
            // Reset all IN_PROGRESS tasks back to PENDING on startup
            await new Promise<void>((resolve, reject) => {
                db.update(
                    { status: TaskStatus.IN_PROGRESS },
                    { $set: { status: TaskStatus.PENDING, started_at: null, finished_at: null, } },
                    { multi: true },
                    (err) => {
                        if (err) {
                            reject(err)
                        } else {
                            resolve()
                        }
                    }
                )
            })
        } catch (error) {
            console.error(error)
        }
    }
    protected async taskWorkerScraperBased(): Promise<void> {
        await this.processScraperBasedTasks();
        setTimeout(this.taskWorkerScraperBased.bind(this), 1000);
    }

    protected async taskWorkerScraperTypeBased(): Promise<void> {
        await this.processScraperTypeBasedTasks();
        setTimeout(this.taskWorkerScraperTypeBased.bind(this), 1000);
    }

    private async processScraperTypeBasedTasks(): Promise<void> {
        const browserScrapers = Server.getBrowserScrapers().length > 0
        const requestScrapers = Server.getRequestScrapers().length > 0
        const taskScrapers = Server.getTaskScrapers().length > 0

        if (browserScrapers) {
            await this.processTasksByType(ScraperType.BROWSER)
        }

        if (requestScrapers) {
            await this.processTasksByType(ScraperType.REQUEST)
        }

        if (taskScrapers) {
            await this.processTasksByType(ScraperType.TASK)
        }
    }

    private async processScraperBasedTasks(): Promise<void> {
        const scrapers = Server.getScrapersNames()
        for (const scraperName of scrapers) {
            await this.processTasksByScraper(scraperName)
        }
    }

    private async processTasksByType(scraperType: string): Promise<void> {
        const [hasCapacity, currentCount, rateLimit]  = this.getCapacityInfo(scraperType)
        if (hasCapacity) {
            await this.executePendingTasks(
                this.buildScraperTypeQuery(scraperType),
                currentCount,
                rateLimit,
                scraperType
            )
        }
    }
    protected buildScraperTypeQuery(scraperType: string): any {
        return {
            $and: [
                { status: TaskStatus.PENDING },
                { scraper_type: scraperType },
                { is_all_task: false },
            ],
        }
    }

    getCapacityInfo(key: string): [boolean, number, number|null] {
        const rateLimit: number | null = this.getMaxRunningCount(key);
        const currentCount: number = this.getCurrentRunningCount(key);
        const hasCapacity: boolean = isNullish(rateLimit) || currentCount < rateLimit;
        return [hasCapacity, currentCount, rateLimit];
    }
    public hasCapacity(key: string): boolean {
        return this.getCapacityInfo(key)[0];
    }
    private async processTasksByScraper(scraperName: string): Promise<void> {
        const [hasCapacity, currentCount, rateLimit]  = this.getCapacityInfo(scraperName)
        if (hasCapacity) {
            await this.executePendingTasks(
                this.buildScraperNameQuery(scraperName),
                currentCount,
                rateLimit,
                scraperName
            )
        }
    }

    protected buildScraperNameQuery(scraperName: string): any {
        return {
            $and: [
                { status: TaskStatus.PENDING },
                { scraper_name: scraperName },
                { is_all_task: false },
            ],
        }
    }

    private getMaxRunningCount(key: string): number {
        // @ts-ignore
        return Server.getRateLimit()[key]
    }

    private getCurrentRunningCount(key: string): number {
        // @ts-ignore
        return this.currentCapacity[key] ?? 0

    }

    private async executePendingTasks(
        taskFilter: any,
        current: number,
        limit: number|null,
        key: string
    ): Promise<void> {
        // @ts-ignore
        const tasks = await getPendingTasks(taskFilter, isNotNullish(limit) ?  limit - current  :null)
        
        for (const task of tasks) {
            this.runTaskAndUpdateCapacity(key, task)
        }
    }

    protected runTaskAndUpdateCapacity(
        key: string,
        taskJson: any
    ): void {
        setImmediate(this.runTask.bind(this), taskJson)
        this.incrementCapacity(key)
    }
     incrementCapacity(key: string): void {
        // @ts-ignore
        this.currentCapacity[key] = (this.currentCapacity[key] ?? 0) + 1;
    }
    
     decrementCapacity(key: string): void {
        // @ts-ignore
        this.currentCapacity[key] = (this.currentCapacity[key] ?? 0) - 1;
    }
    
    /**
     * Adds a task to the abortion check set and starts the shared interval if not running.
     */
    private addToAbortCheck(taskId: number): void {
        this.toCheckAbortTasks.add(taskId)
        this.abortedTasksState[taskId] = false
        
        // Start the shared interval if not already running
        if (!this.abortCheckIntervalId) {
            this.abortCheckIntervalId = setInterval(
                () => this.checkAllTasksAbortion(),
                this.ABORT_CHECK_INTERVAL
            )
        }
    }

    /**
     * Removes a task from the abortion check set and cleans up if no more tasks.
     */
    private removeFromAbortCheck(taskId: number): void {
        this.toCheckAbortTasks.delete(taskId)
        delete this.abortedTasksState[taskId]
        
        // Stop the interval if no more tasks to check
        if (this.toCheckAbortTasks.size === 0 && this.abortCheckIntervalId) {
            clearInterval(this.abortCheckIntervalId)
            this.abortCheckIntervalId = null
        }
    }

    /**
     * Checks abortion status for all tasks in the set.
     * Called by the shared interval.
     */
    private async checkAllTasksAbortion(): Promise<void> {
        if (this.toCheckAbortTasks.size === 0) {
            // No tasks to check, stop the interval
            if (this.abortCheckIntervalId) {
                clearInterval(this.abortCheckIntervalId)
                this.abortCheckIntervalId = null
            }
            return
        }

        try {
            const taskIds = Array.from(this.toCheckAbortTasks)
            const results = await this.getTasksAbortionResults(taskIds)
            
            let allAborted = true
            for (const taskIdStr of Object.keys(results)) {
                const taskId = Number(taskIdStr)
                const isAborted = results[taskId]
                this.abortedTasksState[taskId] = isAborted
                if (isAborted) {
                    // Remove aborted tasks from the check set
                    this.toCheckAbortTasks.delete(taskId)
                } else {
                    allAborted = false
                }
            }
            
            // Stop interval if all tasks are aborted or set is empty
            if (allAborted || this.toCheckAbortTasks.size === 0) {
                if (this.abortCheckIntervalId) {
                    clearInterval(this.abortCheckIntervalId)
                    this.abortCheckIntervalId = null
                }
            }
        } catch (error) {
            console.error('Error checking tasks abortion status:', error)
        }
    }

    /**
     * Batch fetches abortion results for multiple task IDs.
     * Returns an object of taskId -> isAborted.
     */
    async getTasksAbortionResults(taskIds: number[]): Promise<Record<number, boolean>> {
        if (taskIds.length === 0) {
            return {}
        }
        const results: Record<number, boolean> = {}

        // Batch query for all tasks
        const tasks = await db.findAsync(
            { id: { $in: taskIds } },
            { id: 1, status: 1 }
        ) as unknown as Array<{ id: number, status: string }>
        
        const taskIdToTaskMap: Record<number, { id: number, status: string }> = {}
        for (const task of tasks) {
            taskIdToTaskMap[task.id] = task
        }
        
        // Check each task
        for (const taskId of taskIds) {
            if (!taskIdToTaskMap[taskId]) {
                // Task was deleted - treat as aborted
                results[taskId] = true
            } else {
                const task = taskIdToTaskMap[taskId]
                // We do not check for ABORTED status here as even it is failed, failed then restart, we want task to exit
                results[taskId] = task?.status !== TaskStatus.IN_PROGRESS
            }
        }
        
        return results
    }

    /**
     * Creates an isAborted function for a task.
     * Uses centralized abortion checking for efficiency.
     */
    protected createIsAborted(taskId: number): { isAborted: () => boolean, cleanup: () => void } {
        let isFirstCall = true

        const isAborted = () => {
            if (isFirstCall) {
                isFirstCall = false
                // Add to centralized abortion checking
                this.addToAbortCheck(taskId)
            }
            return this.abortedTasksState[taskId] ?? false
        }

        const cleanup = () => {
            this.removeFromAbortCheck(taskId)
        }

        return { isAborted, cleanup }
    }

    protected async runTask(task: any): Promise<void> {
        const key = Server.isScraperBasedRateLimit ? task.scraper_name : task.scraper_type
        const taskId = task.id
        const scraperName = task.scraper_name
        const parent_task_id = task.parent_task_id
        const metadata = isNotEmptyObject(task.metadata) ? { metadata: task.metadata } : {}
        const taskData = task.data

        const fn = Server.getScrapingFunction(scraperName)
        const { isAborted, cleanup } = this.createIsAborted(taskId)
        const removeDuplicatesBy = Server.getRemoveDuplicatesBy(scraperName)
        
        // Create PushDataWriter for pushData functionality
        const onResultCountUpdate = this.createResultCountUpdater(taskId)
        const pushDataWriter = new PushDataWriter(taskId, removeDuplicatesBy, onResultCountUpdate)
        const pushData = pushDataWriter.push.bind(pushDataWriter)

        let exceptionLog: any = null
        try {
            let result: any = null
            try {
                result = await fn(taskData, {
                    ...metadata,
                    isAborted,
                    pushData,
                    parallel: null,
                    cache: false,
                    beep: false,
                    raiseException: true,
                    closeOnCrash: true,
                    output: null,
                    createErrorLogs: false,
                    returnDontCacheAsIs: true,
                })
                let isResultDontCached = false
                if (isDontCache(result)) {
                    isResultDontCached = true
                    result = result.data
                }


                this.decrementCapacity(key)
                cleanup()

                if (pushDataWriter.wasUsed()) {
                    // Push any returned result as well
                    await pushData(result)
                    
                    // Close stream and perform normalization (returns final result count)
                    await pushDataWriter.close()
                    
                    // pushData was used - report success with the normalized result
                    await this.reportTaskSuccessWithPushData(
                        taskId,
                        pushDataWriter.getFilePath(),
                        pushDataWriter.getItemCount(),
                        isResultDontCached,
                        scraperName,
                        taskData,
                        parent_task_id,
                        key
                    )
                } else {
                    // Normal flow - result was returned
                    result = cleanDataInPlace(result)
                    if (removeDuplicatesBy && Array.isArray(result)) {
                        result = removeDuplicatesByKey(result, removeDuplicatesBy)
                    }
                    await this.reportTaskSuccess(taskId, result, isResultDontCached, scraperName, taskData, parent_task_id, key)
                }
                
            } catch (error) {
                await pushDataWriter.close()
                cleanup()
                this.decrementCapacity(key)
                exceptionLog = formatExc(error)
                console.error(error)
                await this.reportTaskFailure(taskId, exceptionLog, parent_task_id, key)
            } 
        } catch (error) {
            await pushDataWriter.close()
            cleanup()
            console.error("Error in run_task", error)
        }
    }

    protected async reportTaskSuccessWithPushData(
        taskId: number,
        taskFilePath:string,
        itemCount: number,
        isResultDontCached: boolean,
        scraperName: string,
        taskData: any,
        parent_task_id: number | null,
        _key: string
    ) {
        await this.markTaskAsSuccessWithFile(
            taskId,
            taskFilePath,
            itemCount,
            isResultDontCached ? false : Server.cache,
            scraperName,
            taskData
        )
        if (parent_task_id) {
            await this.updateParentTaskStreaming(taskId)
        }
    }

    protected async reportTaskFailure(taskId: any, exceptionLog: any, parent_task_id: any, _key: string) {
        await this.markTaskAsFailure(taskId, exceptionLog)
        if (parent_task_id) {
            await this.updateParentTask(taskId, exceptionLog)
        }
    }

    protected async reportTaskSuccess(taskId: any, result: any, isResultDontCached: boolean, scraperName: any, taskData: any, parent_task_id: any, _key: string) {
        await this.markTaskAsSuccess(
            taskId,
            result,
            isResultDontCached ? false : Server.cache,
            scraperName,
            taskData
        )
        if (parent_task_id) {
            await this.updateParentTask(taskId, result)
        }
    }

    protected async updateParentTask(taskId: number, result: any[]) {
        const task = await db.findOneAsync({ id: taskId }) as unknown as Task | null
        if (task && task.parent_task_id) {
            const removeDuplicatesBy = Server.getRemoveDuplicatesBy(task.scraper_name)
            await this.completeParentTaskIfPossible(
                task.parent_task_id,
                removeDuplicatesBy,
                result,
            )
        }
    }

    /**
     * Updates parent task by streaming child results (without loading into memory).
     */
    protected async updateParentTaskStreaming(taskId: number) {
        const task = await db.findOneAsync({ id: taskId }) as unknown as Task | null
        if (task && task.parent_task_id) {
            const removeDuplicatesBy = Server.getRemoveDuplicatesBy(task.scraper_name)
            await this.completeParentTaskIfPossibleStreaming(
                taskId,
                task.parent_task_id,
                removeDuplicatesBy,
            )
        }
    }

    async completeParentTaskIfPossible(
        parentId: number,
        removeDuplicatesBy: string | null,
        result: any[]
    ) {
        const parentTask = await TaskHelper.getTask(
            parentId,
            [TaskStatus.PENDING, TaskStatus.IN_PROGRESS],
        )

        if (parentTask) {
            const is_large: any = parentTask.is_large
            await TaskHelper.updateParentTaskResults(parentId, result, is_large)
            await this.finalizeParentIfAllChildrenDone(parentId, removeDuplicatesBy, is_large)
        }
    }

    /**
     * Streams child task results to parent and completes parent if all children done.
     */
    async completeParentTaskIfPossibleStreaming(
        childId: number,
        parentId: number,
        removeDuplicatesBy: string | null,
    ) {
        const parentTask = await TaskHelper.getTask(
            parentId,
            [TaskStatus.PENDING, TaskStatus.IN_PROGRESS],
        )

        if (parentTask) {
            const is_large: any = parentTask.is_large
            await TaskHelper.streamChildToParentResults(childId, parentId, is_large)
            await this.finalizeParentIfAllChildrenDone(parentId, removeDuplicatesBy, is_large)
        }
    }

    /**
     * Checks if all children are done and finalizes the parent task.
     */
    private async finalizeParentIfAllChildrenDone(
        parentId: number,
        removeDuplicatesBy: string | null,
        is_large: boolean
    ) {
        const isAllChildTasksDone = await TaskHelper.areAllChildTaskDone(parentId)
        if (isAllChildTasksDone) {
            const failedChildrenCount = await TaskHelper.getFailedChildrenCount(parentId)
            const status = failedChildrenCount ? TaskStatus.FAILED : TaskStatus.COMPLETED
            await TaskHelper.readCleanSaveTask(
                parentId,
                removeDuplicatesBy,
                status,
                is_large
            )
        }
    }

    async completeAsMuchAllTaskAsPossible(
        parentId: number,
        removeDuplicatesBy: string | null
    ): Promise<void> {
        const isAllChildTasksDone = await TaskHelper.areAllChildTaskDone(parentId)
        if (isAllChildTasksDone) {
            const failedChildrenCount = await TaskHelper.getFailedChildrenCount(
                parentId
            )
            const status = failedChildrenCount
                ? TaskStatus.FAILED
                : TaskStatus.COMPLETED
            await TaskHelper.collectAndSaveAllTask(
                parentId,
                null,
                removeDuplicatesBy,
                status, 
                false,
            )
        } else {
            await TaskHelper.collectAndSaveAllTask(
                parentId,
                null,
                removeDuplicatesBy,
                TaskStatus.IN_PROGRESS,
                false,
            )
        }
    }

    protected async markTaskAsFailure(
        taskId: number,
        exceptionLog: string
    ): Promise<void> {
        await TaskResults.saveTask(taskId, [exceptionLog])
        const updateResult = await TaskHelper.updateTask(
            taskId,
            {
                status: TaskStatus.FAILED,
                finished_at: new Date(),
            },
            [TaskStatus.IN_PROGRESS],
        )
        if (!updateResult) {
            TaskResults.deleteTask(taskId)
        }
    }

    protected async markTaskAsSuccess(
        taskId: number,
        result: any[],
        cacheTask: boolean,
        scraperName: string,
        data: any
    ): Promise<void> {
        const taskPath = await TaskResults.saveTask(taskId, result)
        const isLarge = isLargeFile(taskPath)
        
        if (cacheTask) {
             await TaskResults.saveCachedTask(scraperName, data, result, isLarge)
        }
        await this.updateTaskAsCompleted(taskId, result.length, isLarge)
    }

    /**
     * Marks a task as success when the result file already exists (pushData case).
     */
    protected async markTaskAsSuccessWithFile(
        taskId: number,
        taskFilePath: string,
        itemCount: number,
        cacheTask: boolean,
        scraperName: string,
        data: any
    ): Promise<void> {
        const isLarge = isLargeFile(taskFilePath)
        
        if (cacheTask) {
            await TaskResults.copyTaskToCachedTask(scraperName, data, itemCount, taskFilePath, isLarge)
        }
        await this.updateTaskAsCompleted(taskId, itemCount, isLarge)
    }

    /**
     * Common method to update task status as completed.
     */
    protected async updateTaskAsCompleted(
        taskId: number,
        resultCount: number,
        isLarge: boolean
    ): Promise<void> {
        const updateResult = await TaskHelper.updateTask(taskId, {
            result_count: resultCount,
            status: TaskStatus.COMPLETED,
            finished_at: new Date(),
            is_large: isLarge
        }, [TaskStatus.IN_PROGRESS])
        if (!updateResult) {
            TaskResults.deleteTask(taskId)
        }
    }

    /**
     * Creates a callback to update task result count during pushData.
     * Override in subclasses to customize or disable updates.
     */
    protected createResultCountUpdater(taskId: number): ((result_count: number) => Promise<any>) | undefined {
        return (result_count: number) => TaskHelper.updateTask(taskId, { result_count })
    }
}

export { TaskExecutor }
